src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -558,91 +487,205 @@ private void updateParticipantMetadata(
-        dataApi.updateEntitiesInEntitySet( entitySetIdMap.get( HAS_ENTITY_SET_NAME ),
-                ImmutableMap.of( hasEntityKeyId, hasEntityData ),
-                UpdateType.PartialReplace );
+        Map<UUID, Set<Object>> hasEntityData = getHasEntity( firstDateTime );
+        EntityKey hasEK = getHasEntityKey( hasEntityData );
+        entitiesByEntityKey.put( hasEK, hasEntityData );
 
-        EntityDataKey dst = new EntityDataKey( entitySetIdMap.get( METADATA_ENTITY_SET_NAME ), metadataEntityKeyId );
-        EntityDataKey edge = new EntityDataKey( entitySetIdMap.get( HAS_ENTITY_SET_NAME ), hasEntityKeyId );
-        EntityDataKey src = new EntityDataKey( participantEntitySetId, participantEntityKeyId );
-        DataEdgeKey dataEdgeKey = new DataEdgeKey( src, dst, edge );
-        dataApi.createEdges( Set.of( dataEdgeKey ) );
+        EntityKey participantEK = getParticipantEntityKey( participantESID, participantId );
 
-        stopwatch.stop();
-        logger.info( "updateParticipantMetadata took {} seconds", stopwatch.elapsed(TimeUnit.SECONDS) );
-        logger.info( "updated participant metadata - study {} participant {}", studyId, participantId );
+        // association: participant  => has => metadata
+        edgesByEntityKey.add( Triple.of( participantEK, hasEK, metadataEK ) );
+
+        return Pair.of( entitiesByEntityKey, edgesByEntityKey );
+    }
+
+    private Map<EntityKey, UUID> getEntityKeyIdMap(
+            DataIntegrationApi integrationApi,
+            Set<Triple<EntityKey, EntityKey, EntityKey>> edgesByEntityKey,
+            Set<EntityKey> entityKeys,
+            UUID participantESID,
+            UUID deviceEKID,
+            UUID participantEKID,
+            String participantId,
+            String deviceId
+    ) {
+        Map<EntityKey, UUID> entityKeyIdMap = Maps.newHashMap();
+
+        Set<EntityKey> orderedEntityKeys = Sets.newLinkedHashSet( entityKeys );
+        edgesByEntityKey.forEach(triple -> {
+            orderedEntityKeys.add( triple.getMiddle() );
+            orderedEntityKeys.add( triple.getLeft() );
+            orderedEntityKeys.add( triple.getRight() );
+        });
+
+        List<UUID> entityKeyIds = integrationApi.getEntityKeyIds( orderedEntityKeys );
+
+        List<EntityKey> entityKeyList = new ArrayList<>( orderedEntityKeys );
+        for ( int i = 0; i < orderedEntityKeys.size(); ++i ) {
+            entityKeyIdMap.put( entityKeyList.get( i ), entityKeyIds.get( i ) );
+        }
+
+        // others
+        EntityKey participantEK = getParticipantEntityKey( participantESID, participantId );
+        entityKeyIdMap.put( participantEK, participantEKID );
+
+        EntityKey deviceEK = getDeviceEntityKey( deviceId );
+        entityKeyIdMap.put( deviceEK, deviceEKID );
+
+        return entityKeyIdMap;
+    }
+
+    // group entities by entity set id
+    private Map<UUID, Map<UUID, Map<UUID, Set<Object>>>> groupEntitiesByEntitySetId(
```

Comment:
when we see these super deeply nested maps like this, I always find it useful to have a comment indicating like
`entitySetId -> entityKeyId -> whatever -> blah -> something` to reduce the amount of code-as-documentation needed to understand what's going on